# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [37]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [38]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

dataset = pd.read_csv('science_data_large.csv')
print(dataset.head(15))
print(dataset.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500

## 2. Splitting the dataset

In [39]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset[["Temperature °C", "Mols KCL"]]
y = dataset["Size nm^3"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.10, random_state = 42)

## 3. Perform a Linear Regression

In [40]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = pd.DataFrame([[500, 600]], columns=["Temperature °C", "Mols KCL"])
predicted_output = model.predict(sample_datapoint)
print(f'Predicted output: {predicted_output}')

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f'Score: {score}')

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print(f'Intercept: {intercept}')
print(f'Coefficients: {coefficients}')

Predicted output: [643298.76699403]
Score: 0.8552472077276096
Intercept: -409391.4795834081
Coefficients: [ 866.14641337 1032.69506649]


Write the linear equation of a slime: (example equation: $E = mc^2$)

$h(x) = -409391.48 + 866.15 x_1 + 1032.70 x_2$

where $x_1$ is Temperature °C and $x_2$ is Mols KCL.

Report on score and explain meaning:

The $R^2$ score quantifies how well the input features explain the variability in the output. In this case, our $R^2$ value is about 0.86, which means that 86% of the variability in size $nm^3$ can be explained by the temperature and moles of KCl. This suggests that the model is quite effective in describing the relationship between the input features and the predicted size.

## 4. Use Cross Validation

In [41]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

model = LinearRegression()
cv = KFold(n_splits = 5, shuffle = True, random_state = 42)

scores = cross_val_score(model, X, y, cv = cv, scoring = 'r2')
print(f'Cross validation R^2 scores: {scores}')
print(f'Mean:', scores.mean())
print(f'Standard Deviation:', scores.std())

# Report on their finding and their significance

Cross validation R^2 scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean: 0.8597294202684644
Standard Deviation: 0.018387737139306387


Write findings here:

The cross validation scores are all about 0.85, which is close to 1. This indicates that the model has a strong predictive capacity, which reflects a robust model performance. Hence, the linear regression approach is suitable for this dataset.

## 5. Using Polynomial Regression

In [42]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

X = dataset[["Temperature °C", "Mols KCL"]]
y = dataset["Size nm^3"]

poly = PolynomialFeatures(degree = 2)
X_poly = poly.fit_transform(X)
model = LinearRegression()
cv = KFold(n_splits = 5, shuffle = True, random_state = 42)

scores = cross_val_score(model, X_poly, y, cv = cv, scoring = 'r2')
print(f'Cross validation scores for polynomial regression: {scores}')

model.fit(X_poly, y)
intercept = model.intercept_
coefficients = model.coef_
print(f'Intercept: {intercept}')
print(f'Coefficients: {coefficients}')

# Report on the metrics and output the resultant equation as you did in Part 3.

Cross validation scores for polynomial regression: [1. 1. 1. 1. 1.]
Intercept: 1.6573350876569748e-05
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23113600e-07 -1.05668128e-11
  2.00000000e+00  2.85714287e-02]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$ h(x) = 1.66 * 10^{-5} + 12 x_1 - 1.23 * 10^{-7} x_2 - 1.06 * 10^{-11} x_1 x_2 + 2 x^2 + 2.86 * 10^{-2} x_2^2 $

where $x_1$ is Temperature °C and $x_2$ is Mols KCL.

Report on the score and interpret:

The scores for all 5 folds are 1.0. This means that the model perfectly predicts the target variable in all k-folds during cross validation. Achieving an $R^2$ score of 1.0 indicates that the polynomial regression model explains 100% of the variance in the target variable for all folds.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [43]:
# Load the dataset. Then train and evaluate the classification models.

ds = pd.read_csv('ckd_feature_subset.csv')
X = ds.drop('Target_ckd', axis = 1)
y = ds['Target_ckd']

cv = KFold(n_splits = 5, shuffle = True, random_state = 42)

results = {}

def evaluate_model(model, X, y):
    scores = cross_val_score(model, X, y, cv = cv, scoring = 'accuracy')
    return scores.mean(), scores.std()

# Logistic Regression
logistic_model = LogisticRegression(max_iter = 1000, random_state = 42)
results['Logistic Regression'] = evaluate_model(logistic_model, X, y)

# Support Vector Machines
svc_model = SVC(random_state = 42)
results['Support Vector Machines'] = evaluate_model(svc_model, X, y)

# k-Nearest Neighbors
knn_model = KNeighborsClassifier()
results['k-Nearest Neighbors'] = evaluate_model(knn_model, X, y)

# Neural Network
nn_model = MLPClassifier(random_state = 42, max_iter = 1000)
results['Neural Networks'] = evaluate_model(nn_model, X, y)

results_ds = pd.DataFrame(results, index = ['Average', 'Std Deviation']).T
print(results_ds)

                          Average  Std Deviation
Logistic Regression      0.856559       0.066269
Support Vector Machines  0.928172       0.047601
k-Nearest Neighbors      0.927957       0.052440
Neural Networks          0.935054       0.040466


## Results and Conclusion for Classification Experiments

The neural network model has the highest average accuracy of 0.94 with the lowest standard deviation of 0.04. This indicates both strong predictive capability and consistent performance across the different folds of the cross-validation.

The support vector machines and k-nearest neighbors models performed very similarly with average accuracies of 0.928172 and 0.927957, respectively. The standard deviations for these models are also low, suggesting reliable performance.

Logistic Regression had the lowest average accuracy (0.856559) with the highest standard deviation (0.066269), indicating that while it captured some patterns in the data, it was the least effective model for this specific classification task.

The results of the table show that Neural Networks has the strongest predictive capacity, followed by support vector machines, k-Nearest Neighbors, and logistic regression. This is because of their respective average percentages and standard deviations as discussed above. 

In [44]:
# Experiments with Neural Network.

def test_neural_networks(configs, X, y):
    results = {}
    for name, config in configs.items():
        model = MLPClassifier(**config, random_state=42, max_iter=1000)
        results[name] = evaluate_model(model, X, y)
    return results

nn_configs = {
    'Config 1': {'hidden_layer_sizes': (10,), 'activation': 'relu'},
    'Config 2': {'hidden_layer_sizes': (50, 25), 'activation': 'tanh'},
    'Config 3': {'hidden_layer_sizes': (100,), 'activation': 'logistic'},
}

nn_results = test_neural_networks(nn_configs, X, y)

nn_results_df = pd.DataFrame(nn_results, index=['Average', 'Std Deviation']).T
print(nn_results_df)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterat

           Average  Std Deviation
Config 1  0.908602       0.052041
Config 2  0.954624       0.032815
Config 3  0.915269       0.048238


/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

Config 2 achieved the highest average accuracy of 0.954624 with a standard deviation of 0.032815. This suggests that this configuration not only performed exceptionally well in terms of prediction accuracy but also exhibited consistency across the different folds of cross-validation.

Config 1 had an average accuracy of 0.908602 and a higher standard deviation (0.052041) compared to Config 2. This indicates that while it performed well, it may have been less stable across the folds.

Config 3 showed an average accuracy of 0.915269, which is slightly lower than Config 1 and much lower than Config 2. The standard deviation of 0.048238 indicates moderate stability.

In conclusion, the results indicate that Configuration 2 stands out as the most effective for achieving high accuracy in predicting chronic kidney disease using neural networks. The architectural choices, specifically relating to hidden layers, the number of neurons, and perhaps the activation functions employed, played a significant role in enhancing the model's performance. Further tuning and experimentation with hyperparameters could yield even better results, reinforcing the importance of configuration in deep learning applications.